cpp/src/io/utilities/file_io_utilities.hpp

```diff
@@ -107,7 +110,10 @@ class cufile_input : public cufile_io_base {
    *
    * @return The number of bytes read
    */
-  virtual size_t read(size_t offset, size_t size, uint8_t* dst, rmm::cuda_stream_view stream) = 0;
+  virtual size_t read(size_t offset, size_t size, uint8_t* dst, rmm::cuda_stream_view stream) 
+  {
+    CUDF_FAIL("these classes that support read must override it.");
+  }
```

Comment:
Changes to this and the other `read` function should be reverted. The issue is caused by `read_async`. We should still require derived datasource classes to implement the `read` members.